client/asset/btc/spv_peer_manager.go

```diff
@@ -0,0 +1,236 @@
+			for _, newPeerResolvedName := range resolvedNames {
+				if peerResolvedName == newPeerResolvedName {
+					return name, true
+				}
+			}
+		}
+	}
+	return "", false
+}
+
+// AddPeer connects to a new peer and stores it in the db.
+func (w *SPVPeerManager) AddPeer(addr string) error {
+	return w.addPeer(addr, added)
+}
+
+func (w *SPVPeerManager) addPeer(addr string, source peerSource) error {
+	w.peersMtx.Lock()
+	defer w.peersMtx.Unlock()
+
+	walletPeer := &walletPeer{source: source}
+	resolvedNames, err := resolvePeerNames(addr)
+	if err != nil {
+		w.log.Errorf("failed to resolve peer name: %v", err)
+	} else {
+		if duplicatePeer, found := w.peerWithResolvedNames(resolvedNames); found {
+			return fmt.Errorf("%s and %s resolve to the same node", duplicatePeer, addr)
+		}
+		walletPeer.resolvedNames = resolvedNames
+	}
+
+	w.peers[addr] = walletPeer
+
+	if source != defaultPeer {
+		err = w.peersDB.Store([]byte(addr), added)
+		if err != nil {
+			w.log.Errorf("failed to store peer in db: %w", err)
+		}
+	}
+
+	return w.cs.AddPeer(addr)
+}
+
+// RemovePeer disconnects from a peer added by the user and removes it from
+// the db.
+func (w *SPVPeerManager) RemovePeer(addr string) error {
+	w.peersMtx.Lock()
+	defer w.peersMtx.Unlock()
+
+	err := w.peersDB.Delete([]byte(addr))
+	if err != nil {
+		w.log.Errorf("failed to delete host from peers DB: %v")
+	} else {
+		delete(w.peers, addr)
+	}
+
+	return w.cs.RemoveNodeByAddr(addr)
+}
+
+// resolvePeerNames returns the IP addresses that the address resolves
+// to.
+func resolvePeerNames(addr string) ([]string, error) {
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	ips, err := net.LookupIP(host)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ips) == 0 {
+		return nil, fmt.Errorf("failed to resolve host: %v", host)
+	}
+
+	names := make([]string, 0, len(ips))
+	for _, ip := range ips {
+		switch {
+		case ip.To4() != nil:
+			names = append(names, fmt.Sprintf("%s:%s", ip.To4(), port))
+		case ip.To16() != nil:
+			names = append(names, fmt.Sprintf("[%s]:%s", ip.To16(), port))
+		}
+	}
+
+	return names, nil
+}
+
+// ConnectToInitialWalletPeers connects to the default peers and the peers
+// that were added by the user and persisted in the db.
+func (w *SPVPeerManager) ConnectToInitialWalletPeers() {
+	for _, peer := range w.defaultPeers {
+		w.addPeer(peer, defaultPeer)
+	}
+
+	_ = w.peersDB.ForEach(func(k []byte, v []byte) error {
+		addr := string(k)
+		err := w.addPeer(addr, added)
+		if err != nil {
+			w.log.Errorf("failed to add peer %s: %v", addr, err)
+		}
+		return nil
+	})
+}
+
+// NewSPVPeerManager creates a new SPVPeerManager.
+func NewSPVPeerManager(cs PeerManagerChainService, defaultPeers []string, dir string, log dex.Logger) (*SPVPeerManager, error) {
+	db, err := kvdb.NewFileDB(filepath.Join(dir, "peers.db"), log.SubLogger("PEERDB"))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create peers db: %w", err)
+	}
```

Comment:
Me neither.  A json.peers file that we  open/write/close the JSON on each call (AddPeer, RemovePeer, Peers) rather than holding it open would be preferable.  If any of those fail, the call was a no-op and it logs an error instead of causing a fatal wallet error.  Also it keeps `OpenSPVWallet` free of one error possibility if `walletConstructor` cannot error.

Reply:
I've updated to use a json file.